R/api.R

```diff
@@ -184,19 +191,45 @@ set_crunch_config <- function(cfg = c(
     options(crunch.httr_config = cfg)
 }
 
-#' @importFrom utils packageVersion
+#' Generate or extend the User-Agent string
+#'
+#' By default, the names and versions of curl, httr, and any attached Crunch
+#' packages are included in the User-Agent request header. You can add to this
+#' using this function.
+#' @param ... Additional character terms to add to the User-Agent string
+#' @return The User-Agent string. Provide this appropriately in requests or set
+#' globally with [set_crunch_config()].
+#' @export
+#' @keywords internal
 #' @importFrom curl curl_version
-crunchUserAgent <- function(x) {
+crunch_user_agent <- function(...) {
     ## Cf. httr:::default_ua
-    versions <- c(
-        libcurl = curl_version()$version,
-        curl = as.character(packageVersion("curl")),
-        httr = as.character(packageVersion("httr")),
-        rcrunch = as.character(packageVersion("crunch"))
+    ## Include versions of any of these packages, if attached
+    pkgs <- ua_packages[ua_packages %in% loadedNamespaces()]
+    ua <- c(
+        # Also include the libcurl version
+        paste0("libcurl/", curl_version()$version),
+        mapply(packageUA, pkgs, names(pkgs)),
+        # And any extra bits provided
+        ...
     )
-    ua <- paste0(names(versions), "/", versions, collapse = " ")
-    if (!missing(x)) ua <- paste(ua, x)
-    return(ua)
+    return(paste(ua, collapse = " "))
+}
+
+ua_packages <- c(
```

Comment:
Do we think we will need to mock/change this? If no (and if there's no other reason) it would be a tiny bit easier to read and not hold in memory / remember where that thing is defined if it were included inside of `crunch_user_agent` above.

Reply:
I put it outside because one way you could extend this list would be by `trace()`ing `crunch_user_agent()` and doing like `ua_packages <- c(ua_packages, crnewpackage="crunewpackage")`. But that's a bit much, really. So if you'd rather it be inside the `crunch_user_agent` function, that's fine.